### PR TITLE
Onboard all object types

### PR DESCRIPTION
**Pull request by yubing**
This PR onboard all object types to validation pipeline.

### Test
`$ python3 -m unittest validate/json_diff_test.py`

```
 (PLAT-45083) ~/migrate $ ./validate_pipeline.sh E20211209153920 E20211209154049
Using the session id: V20211216230340
Start validate-instance_profile
############################# validate-instance_profile #################################
---------------- Compare logs/E20211209153920/instance_profiles.log <-> logs/E20211209154049/instance_profiles.log ---------------------
Object counts 1 <-> 1
No diff found.
validate-instance_profile Completed. Total time taken: 0:00:00.000543
Start validate-users
############################# validate-users #################################
---------------- Compare logs/E20211209153920/users.log <-> logs/E20211209154049/users.log ---------------------
Object counts 26 <-> 27
Diff counts {'MISS_SOURCE': 4, 'VALUE_MISMATCH': 1}
validate-users Completed. Total time taken: 0:00:00.002314
Start validate-groups
############################# validate-groups #################################
Files counts 6 <-> 8
---------------- Compare logs/E20211209153920/groups/tomi-acl-test-group <-> logs/E20211209154049/groups/tomi-acl-test-group ---------------------
Object counts 1 <-> 1
No diff found.
MISS_FILE_SOURCE:
> logs/E20211209153920/groups/aaaa
---------------- Compare logs/E20211209153920/groups/tzars <-> logs/E20211209154049/groups/tzars ---------------------
Object counts 1 <-> 1
No diff found.
---------------- Compare logs/E20211209153920/groups/users <-> logs/E20211209154049/groups/users ---------------------
Object counts 1 <-> 1
Diff counts {'MISS_SOURCE': 1}
---------------- Compare logs/E20211209153920/groups/admins <-> logs/E20211209154049/groups/admins ---------------------
Object counts 1 <-> 1
No diff found.
MISS_FILE_SOURCE:
> logs/E20211209153920/groups/testgroup
---------------- Compare logs/E20211209153920/groups/analysts <-> logs/E20211209154049/groups/analysts ---------------------
Object counts 1 <-> 1
No diff found.
---------------- Compare logs/E20211209153920/groups/field <-> logs/E20211209154049/groups/field ---------------------
Object counts 1 <-> 1
No diff found.
validate-groups Completed. Total time taken: 0:00:00.003024
Start validate-user_dirs
############################# validate-user_dirs #################################
---------------- Compare logs/E20211209153920/user_dirs.log <-> logs/E20211209154049/user_dirs.log ---------------------
Object counts 58 <-> 1286
Diff counts {'MISS_SOURCE': 1233, 'MISS_DESTINATION': 5}
validate-user_dirs Completed. Total time taken: 0:00:00.039042
Start validate-user_workspace
############################# validate-user_workspace #################################
---------------- Compare logs/E20211209153920/user_workspace.log <-> logs/E20211209154049/user_workspace.log ---------------------
Object counts 73 <-> 6544
Diff counts {'MISS_SOURCE': 6471}
validate-user_workspace Completed. Total time taken: 0:00:00.182957
Start validate-libraries
############################# validate-libraries #################################
---------------- Compare logs/E20211209153920/libraries.log <-> logs/E20211209154049/libraries.log ---------------------
Object counts 0 <-> 0
No diff found.
validate-libraries Completed. Total time taken: 0:00:00.000375
Start validate-acl_notebooks
############################# validate-acl_notebooks #################################
---------------- Compare logs/E20211209153920/acl_notebooks.log <-> logs/E20211209154049/acl_notebooks.log ---------------------
Object counts 73 <-> 6544
Diff counts {'MISS_SOURCE': 6565}
validate-acl_notebooks Completed. Total time taken: 0:00:00.285237
Start validate-acl_directories
############################# validate-acl_directories #################################
---------------- Compare logs/E20211209153920/acl_directories.log <-> logs/E20211209154049/acl_directories.log ---------------------
Object counts 58 <-> 1286
Diff counts {'MISS_SOURCE': 1259, 'MISS_DESTINATION': 5}
validate-acl_directories Completed. Total time taken: 0:00:00.060441
Start validate-secrets_scopes
############################# validate-secrets_scopes #################################
Files counts 5 <-> 5
---------------- Compare logs/E20211209153920/secret_scopes/foobar <-> logs/E20211209154049/secret_scopes/foobar ---------------------
Object counts 0 <-> 0
No diff found.
---------------- Compare logs/E20211209153920/secret_scopes/table_acl_tests <-> logs/E20211209154049/secret_scopes/table_acl_tests ---------------------
Object counts 2 <-> 2
No diff found.
---------------- Compare logs/E20211209153920/secret_scopes/tests <-> logs/E20211209154049/secret_scopes/tests ---------------------
Object counts 0 <-> 0
No diff found.
---------------- Compare logs/E20211209153920/secret_scopes/foo_for_all <-> logs/E20211209154049/secret_scopes/foo_for_all ---------------------
Object counts 1 <-> 1
No diff found.
---------------- Compare logs/E20211209153920/secret_scopes/foo <-> logs/E20211209154049/secret_scopes/foo ---------------------
Object counts 2 <-> 2
No diff found.
validate-secrets_scopes Completed. Total time taken: 0:00:00.001826
Start validate-secret_scopes_acls
############################# validate-secret_scopes_acls #################################
---------------- Compare logs/E20211209153920/secret_scopes_acls.log <-> logs/E20211209154049/secret_scopes_acls.log ---------------------
Object counts 5 <-> 5
Diff counts {'MISS_SOURCE': 1}
validate-secret_scopes_acls Completed. Total time taken: 0:00:00.002916
Start validate-clusters
############################# validate-clusters #################################
---------------- Compare logs/E20211209153920/clusters.log <-> logs/E20211209154049/clusters.log ---------------------
Object counts 7 <-> 8
Diff counts {'MISS_SOURCE': 4, 'VALUE_MISMATCH': 3, 'MISS_DESTINATION': 1}
validate-clusters Completed. Total time taken: 0:00:00.001535
Start validate-cluster_policies
############################# validate-cluster_policies #################################
---------------- Compare logs/E20211209153920/cluster_policies.log <-> logs/E20211209154049/cluster_policies.log ---------------------
Object counts 5 <-> 12
Diff counts {'MISS_SOURCE': 7, 'VALUE_MISMATCH': 1}
validate-cluster_policies Completed. Total time taken: 0:00:00.000914
Start validate-acl_clusters
############################# validate-acl_clusters #################################
---------------- Compare logs/E20211209153920/acl_clusters.log <-> logs/E20211209154049/acl_clusters.log ---------------------
Object counts 7 <-> 8
Diff counts {'MISS_SOURCE': 1}
validate-acl_clusters Completed. Total time taken: 0:00:00.000818
Start validate-acl_cluster_policies
############################# validate-acl_cluster_policies #################################
---------------- Compare logs/E20211209153920/acl_cluster_policies.log <-> logs/E20211209154049/acl_cluster_policies.log ---------------------
Object counts 5 <-> 12
Diff counts {'MISS_SOURCE': 7}
validate-acl_cluster_policies Completed. Total time taken: 0:00:00.002566
Start validate-instance_pools
############################# validate-instance_pools #################################
---------------- Compare logs/E20211209153920/instance_pools.log <-> logs/E20211209154049/instance_pools.log ---------------------
Object counts 2 <-> 6
Diff counts {'MISS_SOURCE': 4}
validate-instance_pools Completed. Total time taken: 0:00:00.000903
Start validate-database_details
############################# validate-database_details #################################
---------------- Compare logs/E20211209153920/database_details.log <-> logs/E20211209154049/database_details.log ---------------------
Object counts 8 <-> 8
No diff found.
validate-database_details Completed. Total time taken: 0:00:00.000560
Start validate-success_metastore
############################# validate-success_metastore #################################
---------------- Compare logs/E20211209153920/success_metastore.log <-> logs/E20211209154049/success_metastore.log ---------------------
Object counts 20 <-> 19
Diff counts {'MISS_DESTINATION': 1}
validate-success_metastore Completed. Total time taken: 0:00:00.000908
Start validate-table_acls
############################# validate-table_acls #################################
Files counts 1 <-> 1
---------------- Compare logs/E20211209153920/table_acls/00_table_acls.json <-> logs/E20211209154049/table_acls/00_table_acls.json ---------------------
Object counts 15 <-> 15
No diff found.
validate-table_acls Completed. Total time taken: 0:00:00.000978
############################ Validate notebook artifacts ####################################
 111 files changed
See validation_notebooks.log for details.
############################ Validate metastore tables ######################################
 4 files changed, 12 insertions(+), 3 deletions(-)
See validation_metastores.log for details.
```

### CODE REVIEW

File: validate/json_diff.py [hunk 1]
```diff
@@ -173,6 +169,24 @@ def init_diff_logger(base_dir):
...
+        if primary_key == _HASH_PRIMARY_KEY:
+            return str(data)
+        else:
+            return data.get(primary_key, None)
+    elif isinstance(primary_key, list):
+        for primary_key in primary_key:
+            if primary_key in data:
+                return data[primary_key]
+    elif isinstance(primary_key, dict):
+        assert len(primary_key) == 1, f"Dict-type primary key should only contain one element."
```
Comment (kevinkim9264) [inline validate/json_diff.py:183]: can we also log the primary_key in the case of failure for easier debugging?

File: migration_pipeline.py [hunk 2]
```diff
@@ -214,6 +212,94 @@ def add_dir_diff_task(name, dir_path, config, parents=None):
...
+    add_diff_task("validate-success_metastore", "success_metastore.log", DiffConfig(
+        primary_key="table",
+    ))
+    # metastore/ handled by bash script.
+
+    # MetastoreTableACLExportTask
+    add_dir_diff_task("validate-table_acls", "table_acls", DiffConfig(
+        primary_key="__HASH__",
+        ignored_keys=["ExportTimestamp"]
+    ), ".json")
```
Comment (kevinkim9264) [inline migration_pipeline.py:302]: nit: suffix=

File: migration_pipeline.py [hunk 3]
```diff
@@ -229,3 +315,6 @@ def main():
 
 if __name__ == '__main__':
     main()
+
+
+# -rw-r--r--   1 yubing.yin  staff    20843 Dec 13 11:07 acl_jobs.log
```
Comment (kevinkim9264) [inline migration_pipeline.py:320]: remove?

File: tasks/tasks.py [hunk 4]
```diff
@@ -408,7 +407,7 @@ def diff_files(source, destination, config):
 
 class DiffTask(AbstractTask):
     def __init__(self, name, source, destination, config=None):
-        super().__init__(name)
+        super().__init__(name, wmconstants.WM_VALIDATE, name, False)
```
Comment (kevinkim9264) [inline tasks/tasks.py:410]: nit: skip=False

File: tasks/tasks.py [hunk 5]
```diff
@@ -419,16 +418,22 @@ def run(self):
 
 
 class DirDiffTask(AbstractTask):
-    def __init__(self, name, source, destination, config):
-        super().__init__(name)
+    def __init__(self, name, source, destination, config, suffix=None):
+        super().__init__(name, wmconstants.WM_VALIDATE, name, False)
```
Comment (kevinkim9264) [inline tasks/tasks.py:422]: ditto

**Review by kevinkim9264 — COMMENTED** [review]
Looks good! some nits only

File: validate/json_diff.py (hunk 1 above)
Comment (yubing) [reply validate/json_diff.py:183]: Done.

File: migration_pipeline.py (hunk 2 above)
Comment (yubing) [reply migration_pipeline.py:302]: Done.

File: migration_pipeline.py (hunk 3 above)
Comment (yubing) [reply migration_pipeline.py:320]: Done.

File: migration_pipeline.py [hunk 6]
```diff
@@ -214,6 +212,94 @@ def add_dir_diff_task(name, dir_path, config, parents=None):
...
+                ignored_keys=["DatabricksInstancePoolId", "DatabricksInstanceGroupId"]
+            )
+        }
+    ))
+
+    # NotebookExportTask
+    # Handled by bash script.
+
+    # JobsExportTask
+    # No primary key available for jobs.log and acl_jobs.log.
```
Comment (souravkhandelwal09) [inline migration_pipeline.py:287]: how about `name` field in settings for jobs ?

**Review by souravkhandelwal09 — APPROVED** [review]
LGTM

**Review by kevinkim9264 — APPROVED** [review]

**Pull request merged**